astropy/units/physical.py

```diff
@@ -373,36 +373,36 @@ def _dimensionally_compatible_unit(obj):
         else:
             return NotImplemented
 
-    def _dimensional_analysis(self, other, operation):
+    def _dimensional_analysis(self, other: object, operation) -> Self | NotImplemented:  # ???
         other_unit = self._dimensionally_compatible_unit(other)
         if other_unit is NotImplemented:
             return NotImplemented
         other_unit = _replace_temperatures_with_kelvin(other_unit)
         new_unit = getattr(self._unit, operation)(other_unit)
         return new_unit.physical_type
 
-    def __mul__(self, other):
+    def __mul__(self, other: object) -> Self | NotImplemented:
```

Comment:
The standard is actually to not put `NotImplemented` into the type hint of dunder methods.

Reply:
Bit of a flyby, as I was just looking how the typing "looks": these ones seem rather superfluous; are there not some standard defaults for dunder operations like these?

Reply:
@mhvk — that is a great question that I do not know the answer to!  I *think* that static type checkers like mypy still need type hints in order to check dunder methods, but I'm not sure.

When adding type hints to `plasmapy.particles`, I was able to use autotyping to automatically add type hints for dunder methods, which was pretty straightforward.

Reply:
See comment about `__eq__` - it seems much more logical to me to put here things that allow `__mul__` to succeed.